WalletWasabi/Wallets/SmartBlockProvider.cs

```diff
@@ -0,0 +1,58 @@
+using Microsoft.Extensions.Caching.Memory;
+using NBitcoin;
+using Nito.AsyncEx;
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace WalletWasabi.Wallets
+{
+	/// <summary>
+	/// SmartP2pBlockProvider is a blocks provider that can provide
+	/// blocks from multiple requesters.
+	/// </summary>
+	public class SmartBlockProvider : IBlockProvider
+	{
+		public SmartBlockProvider(IBlockProvider provider, IMemoryCache cache)
+		{
+			InnerBlockProvider = provider;
+			Cache = cache;
+		}
+		
+		private IBlockProvider InnerBlockProvider { get; }
+
+		public IMemoryCache Cache { get; }
+
+		private AsyncLock Lock { get; } = new AsyncLock();
+
+		public Task<Block> GetBlockAsync(uint256 blockHash, CancellationToken cancel)
+		{
+			lock (Lock)
+			{
+				string cacheKey = $"{nameof(SmartBlockProvider):nameof(GetBlockAsync)}:{blockHash}";
+				if (!Cache.TryGetValue(cacheKey, out Task<Block> getBlockTask))
+				{
+					getBlockTask = InnerBlockProvider.GetBlockAsync(blockHash, cancel);
+
+					var cacheEntryOptions = new MemoryCacheEntryOptions()
+						.SetSize(10)
+						.SetSlidingExpiration(TimeSpan.FromSeconds(4))
+						.RegisterPostEvictionCallback(callback: EvictionCallback, state: this);
+
+					// Save data in cache.
+					Cache.Set(cacheKey, getBlockTask, cacheEntryOptions);
+				}
+				return getBlockTask;
+			}
+		}
+
+		private void EvictionCallback(object key, object value, EvictionReason reason, object state)
+		{
+			var task = value as Task<Block>;
+			if (task.IsCompleted)
```

Comment:
You used safe cast here, so you should null check too.

Reply:
@lontivero did you see this?

Reply:
The two classes are really different and has completely different purposes:

The `CachedBlockProvider` has its own responsibility, it is a block provider that can give you a blocks that was previously cached, that's all.

This class is responsible for handling concurrent requests to fetch the same blocks from multiple tasks. Basically this class handles and caches `tasks`. 

This separation of concerns and responsibilities is the result of applying the  
_Single Responsible Principle_ (SOLID principles) what says that we should decompose complex logic in small components that perform only one thing. In this way it is easy to develop, review and test each component separately and also compose behaviors.

Reply:
ACK